### ISSUE TITLE
Correlations in random number stream based on clock seeding

### ISSUE THREAD

**Issue by jtkrogel** [QMCPACK/qmcpack#1068]
I've noticed issues like this before, but the problem became clearer when performing the 1000x runs for #1067.  Of the 1000x runs which should have been independent, fully 30% were carbon copy reproductions of others.

This came about for two reasons:
1) (1/3 of cases) The random offset, and hence the seed, appears to be set by the second (at least on the machine I ran on, so probably true for others).  Since four runs could be executing at any given time in the queue I was in, occasionally a pair or more would start in the same second.  This should have significant implications (statistical bias) for doing QMC in a "high-throughput" fashion on a larger machine.

2) (2/3 of cases) The random offset appeared to have a maximum range of 1024 (it rolled back to zero after this) so the available seeds would repeat every 20 minutes.  Since the runs started over the course of about 45 minutes, some runs submitted much later also had the same seed profile as earlier ones and were identical to them.

In addition to the carbon copy problem, runs having a different, but similar random offset (say 1020 vs 1021) resulted in highly correlated runs.  This effect is shown in the figure below, where the x axis is run number (0 to 999, zoomed in) which corresponds to different offsets, and each line is the value of the local energy from a particular MC block.  As is evident from the figure, several similar offsets lead to similar, but not identical, results.
![rerun_correlation](https://user-images.githubusercontent.com/24547905/45430508-77ab2800-b673-11e8-83b2-f33eb0d6b125.png)

The effect of reduced statistics due to hidden correlations is also directly evident from using the 1000x runs to produce a mean that should be equivalent to a single run 1000x longer.  The deviation between the two results is 2.6 sigma, with a probability of occurrence of 1/100.

Does anyone know where the offset range of 1024 comes from?

What could be done to avoid statistical bias in high-throughput runs?

Do we know that something like this is not already happening across MPI ranks?

**Comment by ye-luo** [comment]
1024 is from
https://github.com/QMCPACK/qmcpack/blob/d80d1b608105ece8362eeea1dc8c069440dca477/src/OhmmsApp/RandomNumberControl.cpp#L80
The following steps are take:
1. The per rank RNG generates 1 number (offset base) and bcast to all MPI ranks (1024 is found here)
2. Generate prime numbers up to nprocs*omp_get_max_threads()+2)+offset. And use the first nprocs ones to reinitialize per rank RNG.
3. Initialize per thread RNG with other prime numbers.

I think 1024 is to restrict the amount of prime numbers generated but not used in the end.

**Comment by ye-luo** [comment]
@jtkrogel Are you running with threads? For the correlation demonstrated in the figure, this is expected. If the offset is different by 1, 15 of 16 per thread RNG is the same in a 16 thread run.

**Comment by jtkrogel** [comment]
Why is this just a problem for testing?  This basically means that batched QMCPACK jobs (not in the qmcpack input way, but separate executions) are subject to statistical bias.  This can be done in testing or production.

**Comment by ye-luo** [comment]
I assume a single run itself is fully decorrelated inside.
In production, I don't do the same simulation multiple times with a different seed. All my calculations use a fixed seed.

**Comment by jtkrogel** [comment]
Yes, but you have control over that.  This is a case where there is no control and the default behavior could probably be improved.

**Comment by prckent** [comment]
Observations:
1. With little effort we could do a much better job of picking seeds in this case, e.g. millisecond resolution timers from std::chrono. 
2. We might consider using a better generator than we currently do, including one that allows cheap "skip ahead" operations. This would make it easier to generate unique streams and we could improve the statistical properties at the same time. See e.g. http://www.pcg-random.org/  for one possibility.
